vllm/hpu/ops.py

```diff
@@ -12,86 +12,49 @@
-        values = torch.zeros((num_seqs, num_query_heads, head_size, block_size), dtype=query.dtype, device="hpu")
-        for seq_id in range(num_seqs):
-            seq_index.fill_(seq_id)
-            if i * block_size < context_lens[seq_id]:
-
-                q =  torch.index_select(query, 0, seq_index).transpose(0, 1)
-                key = torch.index_select(key_cache, 0, block_tables[seq_id][i]).squeeze(0)
-                attn_weight = scale * torch.matmul(q, key)
-
-                if attn_masks is not None:
-                    attn_mask = torch.index_select(attn_masks[i], 0, seq_index)
-                    attn_weight = torch.masked_fill(attn_weight, ~(attn_mask.unsqueeze(0).to(torch.bool)), torch.finfo(attn_weight.dtype).min)
-
-                # FIXME: these dynamic checks serve to ensure the -inf default value is not overwritten with fillers that would cause errors
-                #  in logsoftmax computation. A change to custom block multiplication code is required to avoid incurring extra costs here
-                if context_lens[seq_id] < (i + 1) * block_size:
-                    if context_lens[seq_id] - i*block_size < 0:
-                        attn_weight = torch.finfo(query.dtype).min
-                    else:
-                        attn_weight[:, :, context_lens[seq_id] - i*block_size:] = torch.finfo(query.dtype).min
-                attn_weights.index_copy_(0, seq_index, attn_weight.unsqueeze(0))
-            value = torch.index_select(value_cache, 0, block_tables[seq_id][i])
-            # FIXME: these checks concern filler values in the V cache and should be removed once the underlying issue is addressed
-            value = torch.nan_to_num(value)
-            value[value < -1.0e+30] = 0.0
-            values.index_copy_(0, seq_index, value)
-            torch.hpu.synchronize()
-
-        attn_weights_blocks.append(attn_weights.reshape(num_seqs * num_query_heads, 1, block_size))
-        value_blocks.append(values.reshape(num_seqs * num_query_heads, head_size, block_size).transpose(1, 2))
-
-    exp_sum = torch.zeros((*attn_weights_blocks[0].shape[:2], 1), dtype=attn_weights_blocks[0].dtype, device="hpu")
-    for x in attn_weights_blocks:
-        exp_sum.add_(torch.exp(x).sum(dim=-1, keepdim=True))
-
-    output = torch.zeros_like(query)
-    for i in range(len(attn_weights_blocks)):
-        attention_probs = torch.exp(attn_weights_blocks[i]) / exp_sum
-        value = value_blocks[i]
-        out = torch.matmul(attention_probs.to(value.dtype), value).reshape(num_seqs, num_query_heads, head_size)
-        output.add_(out)
     htorch.core.mark_step()
-    return output.to(dtype=query_in.dtype)
+    batch_size, num_head, head_dim = query_in.shape
+    query_in = query_in.view(batch_size * num_head, 1, head_dim)
+    key = fetch_from_cache(key_cache_in, block_tables)
+    value = fetch_from_cache(value_cache_in, block_tables)
+    seq_len = key.size(-1)
+    attn_weights = torch.bmm(query_in, key).mul_(scale)
```

Comment:
won't we crash here if num query heads != num kv heads?

Reply:
We'd need some additional `repeat_interleave`s for keys and values (i.e. https://github.com/HabanaAI/vllm-fork/pull/4/files#diff-bd4586975c8131c8c80c2103554f75f014ce5d50216126821127a2ea31d96502R255-R259) to handle MQA/GQA cases in LLama70b

Reply:
I'll look into it. Thanks!

Reply:
I've added support for MQA/GQA. Basic unit tests run fine now. Please let me know if anything else is missing :)

Reply:
Looks good, I was able to get Llama 70B serving working properly with your patch and #16.